sscanss/app/widgets/graphics.py

```diff
@@ -298,6 +329,276 @@ def drawBackground(self, painter, rect):
+
+        snapped_anchor = self.grid.snap(self.anchor + self.anchor_offset)
+        diff = snapped_anchor - anchor
+        anchor += diff
+        self.object_moved.emit(diff.x(), diff.y())  # Updates scene transform
+        self.initialized = True
+
+        return anchor
+
+    def snapAnchorToBoxGrid(self, start_pos, stop_pos):
+        """Snaps anchor point to the next box grid intersection in the direction indicated
+        by the mouse drag (start and stop pos)
+
+        :param start_pos: starting scene coordinates
+        :type start_pos: QtCore.QPointF
+        :param stop_pos: stopping scene coordinates
+        :type stop_pos: QtCore.QPointF
+        :return: offset to next grid intersection
+        :rtype: Tuple(float, float)
+        """
+        anchor = self.initialize(self.anchor + self.anchor_offset)
+        strength = 200  # Should find a better way of selecting snap strength
+        diff = QtCore.QPointF(*self.remaining_shift) + (stop_pos - start_pos)
+        shift_amount = (diff.x() // strength, diff.y() // strength)
+        remaining_shift = (diff.x() % strength, diff.y() % strength)
+        dx = shift_amount[0] * self.grid.x
+        dy = shift_amount[1] * self.grid.y
+        new_anchor = QtCore.QPointF(anchor.x() + dx, anchor.y() + dy)
+        self.anchor_offset = new_anchor - self.anchor
+        self.remaining_shift = remaining_shift
+
+        return dx, dy
+
+    def snapAnchorToPolarGrid(self, start_pos, stop_pos, alt):
+        """Snaps anchor point to the next polar grid intersection in the direction indicated
+        by the mouse drag (start and stop pos). The anchor will snap the radial direction only
+        if alt is False otherwise will snap to angular.
+
+        :param start_pos: starting scene coordinates
+        :type start_pos: QtCore.QPointF
+        :param stop_pos: stopping scene coordinates
+        :type stop_pos: QtCore.QPointF
+        :param alt: indicates snap in angular direction
+        :type alt: bool
+        :return: offset to next grid intersection
+        :rtype: Tuple(float, float)
+        """
+        anchor = self.initialize(self.anchor + self.anchor_offset)
+        if alt:
+            adj_stop_pos = stop_pos - self.grid.center
+            adj_start_pos = start_pos - self.grid.center
+
+            if adj_stop_pos.manhattanLength() < 0.1 or adj_start_pos.manhattanLength() < 0.1:
+                return 0, 0
+
+            va = Vector3([adj_start_pos.x(), adj_start_pos.y(), 0.]).normalized
+            vb = Vector3([adj_stop_pos.x(), adj_stop_pos.y(), 0.]).normalized
+            angle = -math.acos(clamp(va | vb, -1.0, 1.0))
+
+            if np.dot([0., 0., 1.], va ^ vb) > 0:
+                angle = -angle
+
+            strength = 0.4  # Should find a better way of selecting snap strength
+            shift_amount = math.trunc((self.remaining_shift[0] + angle) / strength)
+            remaining_shift = math.fmod(self.remaining_shift[0] + angle, strength)
```

Comment:
I think this would do the same and is maybe neater:
`shift_amount, remaining_shift = divmod(self.remaining_shift[0] + angle, strength)`

Same in the other branch of the if

Reply:
The result from divmod is not the same as math.trunc for negative values, for example, divmod(-9, 2) is not the same as math.trunc(-9/2)

Reply:
Ok. I guess in this case that means the difference between always snapping clockwise and snapping towards the starting point.

Reply:
yes